generators/chipyard/src/main/scala/HarnessBinders.scala

```diff
@@ -0,0 +1,279 @@
+    }
+    Nil
+  }
+})
+
+class WithLoopbackNIC extends OverrideHarnessBinder({
+  (system: CanHavePeripheryIceNICModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: NICIOvonly => withClockAndReset(clock, th.harnessReset) {
+        NicLoopback.connect(Some(p), system.p(NICKey))
+      }
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithSimNetwork extends OverrideHarnessBinder({
+  (system: CanHavePeripheryIceNICModuleImp, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val clock = WireInit(false.B.asClock)
+    ports.map {
+      case p: NICIOvonly => SimNetwork.connect(Some(p), clock, th.harnessReset.asBool)
+      case c: Clock => clock := c
+    }
+    Nil
+  }
+})
+
+class WithSimAXIMem extends OverrideHarnessBinder({
+  (system: CanHaveMasterAXI4MemPort, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val p: Parameters = chipyard.iobinders.GetSystemParameters(system)
+    val clock = WireInit(false.B.asClock)
+    ports.filter(_.isInstanceOf[Clock]).map { case p: Clock => clock := p }
+    val axi4_ports = ports.filter(_.isInstanceOf[AXI4Bundle])
+    (axi4_ports zip system.memAXI4Node.edges.in).map { case (port: AXI4Bundle, edge) =>
+      val mem = LazyModule(new SimAXIMem(edge, size=p(ExtMem).get.master.size)(p))
+      withClockAndReset(clock, th.harnessReset) {
+        Module(mem.module).suggestName("mem")
+      }
+      mem.io_axi4.head <> port
+    }
+    Nil
+  }
+})
+
+class WithBlackBoxSimMem extends OverrideHarnessBinder({
+  (system: CanHaveMasterAXI4MemPort, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val p: Parameters = chipyard.iobinders.GetSystemParameters(system)
+    val clock = WireInit(false.B.asClock)
+    ports.filter(_.isInstanceOf[Clock]).map { case p: Clock => clock := p }
+    val axi4_ports = ports.collect { case p: AXI4Bundle => p }
+    (axi4_ports zip system.memAXI4Node.edges.in).map { case (port: AXI4Bundle, edge) =>
+      val memSize = p(ExtMem).get.master.size
+      val lineSize = p(CacheBlockBytes)
+      val mem = Module(new SimDRAM(memSize, lineSize, edge.bundle)).suggestName("simdram")
+      mem.io.axi <> port
+      mem.io.clock := clock
+      mem.io.reset := th.harnessReset
+    }
+    Nil
+  }
+})
+
+class WithSimAXIMMIO extends OverrideHarnessBinder({
+  (system: CanHaveMasterAXI4MMIOPort, th: HasHarnessSignalReferences, ports: Seq[Data]) => {
+    val p: Parameters = chipyard.iobinders.GetSystemParameters(system)
+    val clock = WireInit(false.B.asClock)
+    ports.filter(_.isInstanceOf[Clock]).map { case p: Clock => clock := p }
+    (ports zip system.mmioAXI4Node.edges.in).zipWithIndex.map { case ((port: AXI4Bundle, edge), i) =>
+      val mmio_mem = LazyModule(new SimAXIMem(edge, size = 4096)(p))
+      withClockAndReset(clock, th.harnessReset) {
+        Module(mmio_mem.module).suggestName(s"mmio_mem_${i}")
```

Comment:
I think we should eliminate the zipwithindex and let chisel/firrtl to the postfixing here as we did in the other suggest names for simaximem et al.